marlowe-playground-client/src/Marlowe/Extended.purs

```diff
@@ -534,3 +703,23 @@ instance toCoreContract :: ToCore Contract S.Contract where
   toCore (When cases tim cont) = S.When <$> traverse toCore cases <*> toCore tim <*> toCore cont
   toCore (Let varId val cont) = S.Let <$> pure varId <*> toCore val <*> toCore cont
   toCore (Assert obs cont) = S.Assert <$> toCore obs <*> toCore cont
+
+instance templateContract :: Template Contract Placeholders where
+  getPlaceholderIds Close = mempty
+  getPlaceholderIds (Pay accId payee tok val cont) = getPlaceholderIds val <> getPlaceholderIds cont
+  getPlaceholderIds (If obs cont1 cont2) = getPlaceholderIds obs <> getPlaceholderIds cont1 <> getPlaceholderIds cont2
+  getPlaceholderIds (When cases tim cont) = foldMap getPlaceholderIds cases <> getPlaceholderIds tim <> getPlaceholderIds cont
+  getPlaceholderIds (Let varId val cont) = getPlaceholderIds val <> getPlaceholderIds cont
+  getPlaceholderIds (Assert obs cont) = getPlaceholderIds obs <> getPlaceholderIds cont
+
+instance fillableContract :: Fillable Contract TemplateContent where
+  fillTemplate placeholders contract = case contract of
```

Comment:
why do a `case` here when in the Template implementation you match in the function definition?

Reply:
Because we have an extra parameter so we have a lot of repetition (the go function needs to be inlined everywhere, I think). `when` only affects one clause right?

I preferred using the definition everywhere, but in this case I had to concede because it was so much shorter

Reply:
ah of course.